Add parts types.

fernape opened the pull request:
The input SDL statement will be splitted in different `queryParts`. We support three classes:

 * PassthroughPart
 * InputPart
 * OutputPart

Input and Output parts provide constructors and accessors for convenience. The constructors check for the correctness of the parameters and returns an error if the expressions are malformed.

The definition of the `QueryPart` interface has been moved to the `parts.go` file and has been made private.

juanmanuel-tirado reviewed (approved):
LG2M

fernape committed:
Add parts types.

The input SDL statement will be splitted in different `queryParts`. We
support three classes:

 * PassthroughPart
 * InputPart
 * OutputPart

Input and Output parts provide constructors and accessors for
convenience. The constructors check for the correctness of the
parameters and returns an error if the expressions are malformed.

The definition of the `QueryPart` interface has been moved to the
`parts.go` file and has been made private.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+
+import "fmt"
+
+// queryPart is an interface to be implemented by all parts.
```
This is a redundant definition: _"Foo must be implemented by all foos."._  In such cases it's better to try to convey why this is helpful, etc. For example, something along the lines of:
```
// A queryPart represents a section of a parsed SQL statement, which forms
// a complete query when processed together with its surrounding parts, in
// their correct order.
```

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+
+// queryPart is an interface to be implemented by all parts.
+type queryPart interface {
+	// String-like output for debugging purposes
```
There's a strong convention in Go for how to spell the beginning of method and function documentation:
```
String returns the part's representation for debugging purposes.
```

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+	// String-like output for debugging purposes
+	String() string
+
+	// Translate part to its Sql equivalent
```
Similarly:
```
// ToSQL returns the SQL representation of the part.
```
Let's please use capitalized `SQL` across the board as this is its common form.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+	ToSql() string
+}
+
+// qualifiedExpression represents a qualified expression
```
Same as above, "fooBar represents a foo bar" is not helpful and it wastes the reader time without adding additional knowledge (if there's text we tend to read it regardless). Besides addressing the docs, though, maybe we can also improve a bit the naming here, to help people figure what this is, since neither "qualifier" nor "expression" are very indicative of what it really does for us (e.g. "1+1" is also an expression, but that's not it).

How about:
```
// fullName represents a column or field name together with its Go type or table name. 
type fullName struct {
        Prefix string
        Name string
}
```

Or perhaps, even better, can we split it up?
```
type fieldName struct {
        Type, Name string
}
type columnName struct {
        Table, Name string
}
```

fernape replied:
Using `FullName` so the type is exported.

fernape replied:
The thing here was to use `qualifier` as a very common name in grammars, especially SQL ones:
https://go.dev/ref/spec#Qualified_identifiers
http://www.contrib.andrew.cmu.edu/~shadow/sql/sql1992.txt
```
    <qualified name> ::=
              [ <schema name> <period> ] <qualified identifier>

niemeyer replied:
I understand the intention, but have you noticed how the very documentation for this type was explaining what "qualified" and "unqualified" meant for the particular context?

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+// qualifiedExpression represents a qualified expression
+// qualified. Examples are:
+//	... column as &Person ... 	<--- "column" is unqualified
+//	... t.column as &Person.name	<--- "name" and "column" and qualified
```
The examples were useful, but it's a symptom of the issues described above. In other words, if we get the type name and documentation to be clear, it becomes obvious and we don't need the examples anymore.

fernape replied:
Removed since we changed the naming.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+	name      string
+}
+
+// InputPart represents an expression used as input in the SDL query.
```
```
// InputPart represents a named parameter that will be sent to the database to
// while performing the query.
```

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+				name: tagName,
+			},
+		},
+	}, nil
```
That's quite a bit of syntax to replace what seems trivial:
```
&InputPart{
        Type: "Foo",
        Name: "bar",
}
```
Why the additional complexity? We don't need the error checking here as we can do it when processing the query.

Also, this is smelling slightly a bit like a name soup, in the sense we have several different ways to refer to the same data. For example, a Go field is now a "name", a "tag", and an "expression", that sits in a "type", or a "qualifier". We'll likely benefit from simplifying this: it's a type name, with a field name.

just here we have three different ways to name the same thing: tagName,

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+	}, nil
+}
+
+func (ip *InputPart) String() string {
```
Nitpick: "ip" reads as "IP" :)

I suggest using just "p" for all parts.

fernape replied:
Yep, you're right.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+// TagName returns the name of the type in the output part or empty string if
+// none is present.
+func (gt *goType) TagName() string {
+	return gt.name
```
I'm a bit lost by now. There's a lot of abstraction in the types and methods above that just feels like renaming data back and forth. I may be missing part of the picture, but I'm really hoping that we can dramatically simplify this since all of that back and forth is simply handling (type, name) with half a dozen different names.

fernape replied:
Done.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+
+// passthroughPart represents a part of the SDL that we want to pass to the
+// backend database verbatim.
+type PassthroughPart struct {
```
"Bypass" is a shorter synonym for "Passthrough". Might be nicer to code with.

fernape replied:
Shorted and equally clear. Changed.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+// 	&Person
+//	&Person.name
+type OutputPart struct {
+	Column column
```
For example, considering the point above:
```
type OutputPart struct {
        Source fullName
        Target fullName
}
```
This takes down two or three levels of indirection, and avoids a large number of methods.

fernape replied:
Done, using `FullName`

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,155 @@
+//	&Person.name
+type OutputPart struct {
+	Column column
+	GoType goType
```
There's an additional issue that I'll mention here for the lack of a better place: it's not a good practice to have public fields with private types, as that makes them extremely inconvenient to work with. So we need to review what types from this package are supposed to be public, and in which cases the type data is an implementation detail or publicly exposed data.

fernape replied:
Now the type is public.

The general idea here was to use the same underlying type for expressing input parts and output parts. The latter have columns and go types but the structure is identical: `something` `.` `something`. So we had the underlying `qualifierExpression` that was accessed as `TypeName()` or `TableName()` depending on the location: the former makes sense for go types, the latter only for column specifications. Types here were private (`column` and `goType`) since they had accessor methods. I didn't see any use cases for client code importing the `parse` package and wanting to deal with the inner structure of the parts.

Thanks for the review!

niemeyer reviewed (changes requested):
Thanks, Fernando. Some early comments below. As you can see, quite a bit of this is actually about agreeing on basic conventions and patterns to use. These early few reviews will be a bit more extensive, even if the PR is small, but as we get past that early phase things will become easier and simpler to get in.

fernape committed:
Address feedback from PR #7

 * Use `FullName` to represent identifiers, either column names or Go
   types.
 * Make the type public.
 * Remove accessors.
 * Eliminate indirection levels.
 * s/passthroughPart/BypassPart
 * Use `p` for all parts when used as receivers.

Aflynn50 commented on internal/parse/parts.go:
```diff
@@ -1,155 +1,81 @@
 }
 
 // NewOutputPart is a constructor that returns a pointer to an OutputPart.
 func NewOutputPart(tableName string, colName string,
```
Same as for NewInputPart

fernape replied:
See below

Aflynn50 commented on internal/parse/parts.go:
```diff
@@ -1,155 +1,81 @@
 }
 
 // NewInputPart is a constructor that returns a pointer to an InputPart.
 func NewInputPart(typeName string, tagName string) (*InputPart, error) {
```
These constructors are not as vital now we have a simplified structure, a new InputPart can be instantiated straight from a FullName, which in all likely-hood is what we will have in the natural flow of the parser

fernape replied:
Done.
Removed constructors. Simplified tests.

fernape committed:
Remove constructors for Input/Output parts.

Simplify tests since we don't use constructors anymore and hence we
don't need to check for error whatsoever.

fernape committed:
Add comments to exported methods.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,68 @@
+	Prefix, Name string
+}
+
+// InputPart represents a named parameter that will be send to the database
```
s/send/sent/

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,68 @@
+// InputPart represents a named parameter that will be send to the database
+// while performing the query.
+type InputPart struct {
+	FullName
```
I'm lacking a bit of context on usage, but gut feeling is that I'd make this symmetrical with OutputPart, having a field name here instead of embedding. Maybe Source?

fernape replied:
Done. `Source` and `Target` make more sense for the `OutputPart` but it is not confusing here either, so I suppose we can use it for symmetry.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,68 @@
+	FullName
+}
+
+// String returns a textual representation of the InputPart meant for debugging
```
We don't need these boilerplate docstrings on every one of these methods, as this will be pretty much a copy & paste from the queryPart interface.  As a guideline, we should only document something when the text is actually improving the understanding of the reader about the code at hand. Otherwise, instead of helping we're rather hindering the understanding of the code by forcing the reading of something obvious.

fernape replied:
Removed comments from methods already in the interface.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,68 @@
+	return ""
+}
+
+// ToSQL returns a string with the SQL translation for the InputPart.
```
Clear example of the above: "ToSQL returns a SQL for this Type"

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,68 @@
+	return ""
+}
+
+// OutputPart represents an expression to be used as output in our SDL.
```
This is one case where we can actually help the next person:
```
// OutputPart represents a named target output variable in the SQL expression, as well
// as the source table and column where it will be read from.
```
Again, this would help the person fill the gaps that are not obvious from the code below, rather than simply stating "Output part is an output part".

fernape replied:
Done. It is clearer now.

niemeyer commented on internal/parse/parts.go:
```diff
@@ -0,0 +1,68 @@
+}
+
+// BypassPart represents a part of the SDL that we want to pass to the
+// backend database verbatim.
```
That's nice, but let's please replace SDL with "expression" in our documentation across the board to avoid unnecessary acronyms.

niemeyer reviewed (approved):
This is a lot better, thank you!

Last past with tiny nitpicks, and this is good to go.

fernape committed:
s/send/sent

fernape committed:
Make Input and Output parts symmetric

Use a named parameter instead of embedding.

fernape committed:
Remove superfluous comments

fernape committed:
Remove more superfluous comments

fernape committed:
Improve comments for Input and Bypass parts
